src/Workspaces/Core/Portable/Workspace/Solution/CachedSkeletonReferences.cs

```diff
@@ -4,246 +4,407 @@
+            referenceSet = s_compilationToReferenceMap.GetValue(compilation, _ => referenceSet);
+
+            lock (_stateGate)
             {
-                workspace.LogTestMessage($"We failed to create metadata so we're using the one we just found from an earlier version.");
-                return referenceSet;
+                // whoever won, still store this reference set against us with the provided version.
+                _version = version;
+                _skeletonReferenceSet = referenceSet;
             }
-        }
 
-        return new SkeletonReferenceSet(image, new DeferredDocumentationProvider(finalCompilation));
-    }
+            return referenceSet.GetMetadataReference(properties);
+        }
 
-    /// <summary>
-    /// Tries to get the <see cref="SkeletonReferenceSet"/> for this project matching <paramref name="version"/>.
-    /// if <paramref name="version"/> is <see langword="null"/>, any cached <see cref="SkeletonReferenceSet"/> 
-    /// can be returned, even if it doesn't correspond to that version.  This is useful in error tolerance cases
-    /// as building a skeleton assembly may easily fail. In that case it's better to use the last successfully 
-    /// built skeleton than just have no semantic information for that project at all.
-    /// </summary>
-    private SkeletonReferenceSet? TryGetReferenceSet(VersionStamp? version)
-    {
-        // Otherwise, we don't have a direct mapping stored.  Try to see if the cached reference we have is
-        // applicable to this project semantic version.
-        lock (_gate)
+        private async Task<SkeletonReferenceSet> GetOrBuildReferenceSetAsync(
+            Workspace workspace,
+            VersionStamp version,
+            Compilation compilation,
+            CancellationToken cancellationToken)
         {
-            // if we don't have a skeleton cached, then we have nothing to return.
-            if (_skeletonReferenceSet == null)
-                return null;
+            var referenceSet = TryGetExistingReferenceSet(version, compilation);
+            if (referenceSet != null)
+                return referenceSet;
 
-            // if the caller is requiring a particular semantic version, it much match what we have cached.
-            if (version != null && version != _version)
-                return null;
+            // okay, we don't have one. so create one now.
+            ITemporaryStreamStorage? storage;
 
-            return _skeletonReferenceSet;
+            using (await _emitGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // after taking the gate, another thread may have succeeded.  See if we can use their version if so:
+                referenceSet = TryGetExistingReferenceSet(version, compilation);
+                if (referenceSet != null)
+                    return referenceSet;
+
+                storage = TryCreateMetadataStorage(workspace, compilation, cancellationToken);
```

Comment:
I think there may be a race here?
1.  threadA hits line 163, enters the semaphore and starts creating the storage
2.  threadB hits line 163 while 1) is running and starts waiting
3.  threadA finishes creating the storage, exits the semaphore, but doesn't yet return the new skeleton references / update the fields
4.  threadB enters the semaphore, doesn't see the result from 3) since it isn't updated, and starts creating the same storage again. 

Maybe the updating of the locals needs to happen in the semaphore as well